panic: interface conversion: dao.NativeContractCache is nil, not *native.PolicyCache

```
% ./neo-go-linux-amd64 node --config-file config.yml
2023-10-31T20:13:12.354+0300	INFO	P2PNotaryRequestPayloadPool size is not set or wrong, setting default value	{"P2PNotaryRequestPayloadPoolSize": 1000}
2023-10-31T20:13:12.354+0300	INFO	MaxBlockSize is not set or wrong, setting default value	{"MaxBlockSize": 262144}
2023-10-31T20:13:12.354+0300	INFO	MaxBlockSystemFee is not set or wrong, setting default value	{"MaxBlockSystemFee": 900000000000}
2023-10-31T20:13:12.354+0300	INFO	MaxTransactionsPerBlock is not set or wrong, using default value	{"MaxTransactionsPerBlock": 512}
2023-10-31T20:13:12.354+0300	INFO	MaxValidUntilBlockIncrement is not set or wrong, using default value	{"MaxValidUntilBlockIncrement": 5760}
2023-10-31T20:13:12.354+0300	INFO	Hardforks are not set, using default value
2023-10-31T20:13:12.356+0300	INFO	restoring blockchain	{"version": "0.2.10"}
panic: interface conversion: dao.NativeContractCache is nil, not *native.PolicyCache

goroutine 1 [running]:
github.com/nspcc-dev/neo-go/pkg/core/native.(*Policy).isBlockedInternal(0xc000f97340?, 0xc000fff080?, {0x4e, 0x7d, 0x51, 0x11, 0xe7, 0x22, 0x2f, 0xd8, ...})
	github.com/nspcc-dev/neo-go/pkg/core/native/policy.go:258 +0x149
github.com/nspcc-dev/neo-go/pkg/core/native.(*Policy).IsBlocked(...)
	github.com/nspcc-dev/neo-go/pkg/core/native/policy.go:250
github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).getCandidates.func1({0xc000f55596, 0x21, 0x21}, {0xc001002130, 0xa, 0xa})
	github.com/nspcc-dev/neo-go/pkg/core/native/native_neo.go:981 +0x1d1
github.com/nspcc-dev/neo-go/pkg/core/dao.(*Simple).Seek.func1({0xc000f55590?, 0x0?, 0x0?}, {0xc001002130?, 0x10d6ac0?, 0x0?})
	github.com/nspcc-dev/neo-go/pkg/core/dao/dao.go:387 +0x45
github.com/nspcc-dev/neo-go/pkg/core/storage.performSeek.func3({0x7fb075b6576d, 0x27, 0x410725?}, {0x7fb075b65794, 0xa, 0x1?})
	github.com/nspcc-dev/neo-go/pkg/core/storage/memcached_store.go:294 +0x46a
github.com/nspcc-dev/neo-go/pkg/core/storage.(*BoltDBStore).Seek.func1(0xc000ff7ae0?, {0x7fb075b6576d?, 0x2?, 0x0?}, {0x7fb075b65794?, 0x3fe0000000000000?, 0xc00037a168?})
	github.com/nspcc-dev/neo-go/pkg/core/storage/boltdb_store.go:131 +0x34
github.com/nspcc-dev/neo-go/pkg/core/storage.boltSeek.func1(0x60?)
	github.com/nspcc-dev/neo-go/pkg/core/storage/boltdb_store.go:163 +0x2f0
go.etcd.io/bbolt.(*DB).View(0x10edf20?, 0xc000ff0480)
	go.etcd.io/bbolt@v1.3.7/db.go:893 +0x72
github.com/nspcc-dev/neo-go/pkg/core/storage.boltSeek(0xc00037a2c8, {{0xc001002060, 0x6, 0x6}, {0x0, 0x0, 0x0}, 0x0, 0x0}, 0xc000125730)
	github.com/nspcc-dev/neo-go/pkg/core/storage/boltdb_store.go:140 +0x13a
github.com/nspcc-dev/neo-go/pkg/core/storage.(*BoltDBStore).Seek(0x1d95580?, {{0xc001002060, 0x6, 0x6}, {0x0, 0x0, 0x0}, 0x0, 0x0}, 0xc0002cebd0)
	github.com/nspcc-dev/neo-go/pkg/core/storage/boltdb_store.go:130 +0xf2
github.com/nspcc-dev/neo-go/pkg/core/storage.performSeek({0x148fd78, 0x1d929a0}, {0x1491018, 0xc00011a008}, {0x0, 0x0, 0x0}, {{0xc001002060, 0x6, 0x6}, ...}, ...)
	github.com/nspcc-dev/neo-go/pkg/core/storage/memcached_store.go:308 +0x475
github.com/nspcc-dev/neo-go/pkg/core/storage.(*MemCachedStore).Seek(0x0?, {{0xc001002060, 0x6, 0x6}, {0x0, 0x0, 0x0}, 0x0, 0x0}, 0xc000ffa640)
	github.com/nspcc-dev/neo-go/pkg/core/storage/memcached_store.go:157 +0xe8
github.com/nspcc-dev/neo-go/pkg/core/dao.(*Simple).Seek(0xc000172400, 0xfffffffb, {{0xc001002060, 0x6, 0x6}, {0x0, 0x0, 0x0}, 0x0, 0x0}, ...)
	github.com/nspcc-dev/neo-go/pkg/core/dao/dao.go:386 +0x2b2
github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).getCandidates(0xc0001ea2c0, 0xc000172400, 0x0, 0xffffffffffffffff)
	github.com/nspcc-dev/neo-go/pkg/core/native/native_neo.go:978 +0x1e9
github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).computeCommitteeMembers(0xc0001ea2c0, 0x563a7, 0xc000fe7eb0?)
	github.com/nspcc-dev/neo-go/pkg/core/native/native_neo.go:1192 +0x1cd
github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).updateCachedNewEpochValues(0xc0001ea2c0?, 0xc000f4fd40?, 0xc000f4fd40, 0x1?, 0x7)
	github.com/nspcc-dev/neo-go/pkg/core/native/native_neo.go:399 +0x2f
github.com/nspcc-dev/neo-go/pkg/core/native.(*NEO).InitializeCache(0xc0001ea2c0, 0x563a7, 0x1f5ea5be9dd67767?)
	github.com/nspcc-dev/neo-go/pkg/core/native/native_neo.go:351 +0x48a
github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).initializeNativeCache(0xc000172c80?, 0x563a7, 0x1?)
	github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:1025 +0xac
github.com/nspcc-dev/neo-go/pkg/core.(*Blockchain).init(0xc000005500)
	github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:475 +0x62e
github.com/nspcc-dev/neo-go/pkg/core.NewBlockchain({0x1491018?, 0xc00011a008}, {{0x0, {0x0, 0x0}, 0x2bdb2b5f, 0xc350, 0xc0003c6510, 0x12795f58d50000, 0x3e8, ...}, ...}, ...)
	github.com/nspcc-dev/neo-go/pkg/core/blockchain.go:331 +0x1898
github.com/nspcc-dev/neo-go/cli/server.initBlockChain({{0x0, {0x0, 0x0}, 0x2bdb2b5f, 0xc350, 0x0, 0x12795f58d50000, 0x0, 0x0, 0x0, ...}, ...}, ...)
	github.com/nspcc-dev/neo-go/cli/server/server.go:656 +0x207
github.com/nspcc-dev/neo-go/cli/server.initBCWithMetrics({{0x0, {0x0, 0x0}, 0x2bdb2b5f, 0xc350, 0x0, 0x12795f58d50000, 0x0, 0x0, 0x0, ...}, ...}, ...)
	github.com/nspcc-dev/neo-go/cli/server/server.go:135 +0x58
github.com/nspcc-dev/neo-go/cli/server.startServer(0xc00017a840)
	github.com/nspcc-dev/neo-go/cli/server/server.go:464 +0x545
github.com/urfave/cli.HandleAction({0xff7440?, 0x13339f0?}, 0x4?)
	github.com/urfave/cli@v1.22.5/app.go:524 +0x50
github.com/urfave/cli.Command.Run({{0x11dc19a, 0x4}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x11ebab4, 0x12}, {0x1223796, ...}, ...}, ...)
	github.com/urfave/cli@v1.22.5/command.go:173 +0x63e
github.com/urfave/cli.(*App).Run(0xc0001716c0, {0xc00003e080, 0x4, 0x4})
	github.com/urfave/cli@v1.22.5/app.go:277 +0xb07
main.main()
	./main.go:12 +0x2d
zsh: exit 2     ./neo-go-linux-amd64 node --config-file config.yml
```

steps to reproduce:
1. start neogo with no database
2. interrupt with ctrl-c, it will say:
```
^C2023-10-31T20:11:54.815+0300	INFO	shutting down server	{"peers": 11}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "170.64.152.72:51333", "error": "server shutdown", "peerCount": 10}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "65.108.90.74:51333", "error": "server shutdown", "peerCount": 9}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "159.203.28.78:51333", "error": "server shutdown", "peerCount": 8}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "49.12.226.190:50333", "error": "server shutdown", "peerCount": 7}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "5.78.81.79:51333", "error": "server shutdown", "peerCount": 6}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "49.12.226.190:51333", "error": "server shutdown", "peerCount": 5}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "170.64.152.72:50333", "error": "server shutdown", "peerCount": 4}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "116.203.249.217:50333", "error": "server shutdown", "peerCount": 3}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "167.71.212.120:50333", "error": "server shutdown", "peerCount": 2}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "159.203.28.78:50333", "error": "server shutdown", "peerCount": 1}
2023-10-31T20:11:54.815+0300	WARN	peer disconnected	{"addr": "65.108.90.74:50333", "error": "server shutdown", "peerCount": 0}
2023-10-31T20:11:55.201+0300	INFO	persisted to disk	{"blocks": 1367, "keys": 22958, "headerHeight": 353191, "blockHeight": 353191, "took": "385.464896ms"}
```
3. start it again
4. repeat untill subj

--- a/pkg/core/blockchain.go
+++ b/pkg/core/blockchain.go
@@ -468,6 +468,8 @@ func (bc *Blockchain) init() error {
 	}
 	bc.blockHeight = bHeight
 	bc.persistedHeight = bHeight
+
+	bc.log.Debug("initializing caches", zap.Uint32("blockHeight", bHeight))
 	if err = bc.stateRoot.Init(bHeight); err != nil {
 		return fmt.Errorf("can't init MPT at height %d: %w", bHeight, err)
 	}

--- a/pkg/core/blockchain_neotest_test.go
+++ b/pkg/core/blockchain_neotest_test.go
@@ -301,6 +301,42 @@ func TestBlockchain_StartFromExistingDB(t *testing.T) {
 	})
 }
 
+// TestBlockchain_InitializeNeoCache_Bug3181 is aimed to reproduce and check situation
+// when panic occures on native Neo cache initialization due to access to native Policy
+// cache when it's not yet initialized to recalculate candidates.
+func TestBlockchain_InitializeNeoCache_Bug3181(t *testing.T) {
+	ps, path := newLevelDBForTestingWithPath(t, "")
+	bc, validators, committee, err := chain.NewMultiWithCustomConfigAndStoreNoCheck(t, nil, ps)
+	require.NoError(t, err)
+	go bc.Run()
+	e := neotest.NewExecutor(t, bc, validators, committee)
+
+	// Add at least one registered candidate to enable candidates Policy check.
+	acc := e.NewAccount(t, 10000_0000_0000) // block #1
+	neo := e.NewInvoker(e.NativeHash(t, nativenames.Neo), acc)
+	neo.Invoke(t, true, "registerCandidate", acc.(neotest.SingleSigner).Account().PublicKey().Bytes()) // block #2
+
+	// Put some empty blocks to reach N-1 block height, so that newEpoch cached
+	// values of native Neo contract require an update on the subsequent cache
+	// initialization.
+	for i := 0; i < len(bc.GetConfig().StandbyCommittee)-1-2; i++ {
+		e.AddNewBlock(t)
+	}
+	bc.Close() // Ensure persist is done and persistent store is properly closed.
+
+	ps, _ = newLevelDBForTestingWithPath(t, path)
+	t.Cleanup(func() { require.NoError(t, ps.Close()) })
+
+	// Start chain from the existing database that should trigger an update of native
+	// Neo newEpoch* cached values during initializaition. This update requires candidates
+	// list recalculation and policies checks, thus, access to native Policy cache
+	// that is not yet initialized by that moment.
+	require.NotPanics(t, func() {
+		_, _, _, err = chain.NewMultiWithCustomConfigAndStoreNoCheck(t, nil, ps)
+		require.NoError(t, err)
+	})
+}
+
 // This test enables Notary native contract at non-zero height and checks that no
 // Notary cache initialization is performed before that height on node restart.
 func TestBlockchain_InitializeNativeCacheWrtNativeActivations(t *testing.T) {

--- a/pkg/core/dao/dao.go
+++ b/pkg/core/dao/dao.go
@@ -970,7 +970,7 @@ func (dao *Simple) GetRWCache(id int32) NativeContractCache {
 func (dao *Simple) getCache(k int32, ro bool) NativeContractCache {
 	if itm, ok := dao.nativeCache[k]; ok {
 		// Don't need to create itm copy, because its value was already copied
-		// the first time it was retrieved from loser ps.
+		// the first time it was retrieved from lower ps.
 		return itm
 	}
 

--- a/pkg/core/native/policy.go
+++ b/pkg/core/native/policy.go
@@ -365,8 +365,9 @@ func (p *Policy) unblockAccount(ic *interop.Context, args []stackitem.Item) stac
 // like not being signed by a blocked account or not exceeding the block-level system
 // fee limit.
 func (p *Policy) CheckPolicy(d *dao.Simple, tx *transaction.Transaction) error {
+	cache := d.GetROCache(p.ID).(*PolicyCache)
 	for _, signer := range tx.Signers {
-		if _, isBlocked := p.isBlockedInternal(d.GetROCache(p.ID).(*PolicyCache), signer.Account); isBlocked {
+		if _, isBlocked := p.isBlockedInternal(cache, signer.Account); isBlocked {
 			return fmt.Errorf("account %s is blocked", signer.Account.StringLE())
 		}
 	}

--- a/pkg/core/native/policy.go
+++ b/pkg/core/native/policy.go
@@ -241,26 +241,33 @@ func (p *Policy) setExecFeeFactor(ic *interop.Context, args []stackitem.Item) st
 // isBlocked is Policy contract method that checks whether provided account is blocked.
 func (p *Policy) isBlocked(ic *interop.Context, args []stackitem.Item) stackitem.Item {
 	hash := toUint160(args[0])
-	_, blocked := p.isBlockedInternal(ic.DAO, hash)
+	_, blocked := p.isBlockedInternal(ic.DAO.GetROCache(p.ID).(*PolicyCache), hash)
 	return stackitem.NewBool(blocked)
 }
 
-// IsBlocked checks whether provided account is blocked.
+// IsBlocked checks whether provided account is blocked. Normally it uses Policy
+// cache, falling back to the DB queries when Policy cache is not available yet
+// (the only case is native cache initialization pipeline, where native Neo cache
+// is being initialized before the Policy's one).
 func (p *Policy) IsBlocked(dao *dao.Simple, hash util.Uint160) bool {
-	_, isBlocked := p.isBlockedInternal(dao, hash)
+	cache := dao.GetROCache(p.ID)
+	if cache == nil {
+		key := append([]byte{blockedAccountPrefix}, hash.BytesBE()...)
+		return dao.GetStorageItem(p.ID, key) == nil
+	}
+	_, isBlocked := p.isBlockedInternal(cache.(*PolicyCache), hash)
 	return isBlocked
 }
 
 // isBlockedInternal checks whether provided account is blocked. It returns position
 // of the blocked account in the blocked accounts list (or the position it should be
 // put at).
-func (p *Policy) isBlockedInternal(dao *dao.Simple, hash util.Uint160) (int, bool) {
-	cache := dao.GetROCache(p.ID).(*PolicyCache)
-	length := len(cache.blockedAccounts)
+func (p *Policy) isBlockedInternal(roCache *PolicyCache, hash util.Uint160) (int, bool) {
+	length := len(roCache.blockedAccounts)
 	i := sort.Search(length, func(i int) bool {
-		return !cache.blockedAccounts[i].Less(hash)
+		return !roCache.blockedAccounts[i].Less(hash)
 	})
-	if length != 0 && i != length && cache.blockedAccounts[i].Equals(hash) {
+	if length != 0 && i != length && roCache.blockedAccounts[i].Equals(hash) {
 		return i, true
 	}
 	return i, false
@@ -320,7 +327,7 @@ func (p *Policy) blockAccount(ic *interop.Context, args []stackitem.Item) stacki
 	return stackitem.NewBool(p.blockAccountInternal(ic.DAO, hash))
 }
 func (p *Policy) blockAccountInternal(d *dao.Simple, hash util.Uint160) bool {
-	i, blocked := p.isBlockedInternal(d, hash)
+	i, blocked := p.isBlockedInternal(d.GetROCache(p.ID).(*PolicyCache), hash)
 	if blocked {
 		return false
 	}
@@ -343,7 +350,7 @@ func (p *Policy) unblockAccount(ic *interop.Context, args []stackitem.Item) stac
 		panic("invalid committee signature")
 	}
 	hash := toUint160(args[0])
-	i, blocked := p.isBlockedInternal(ic.DAO, hash)
+	i, blocked := p.isBlockedInternal(ic.DAO.GetROCache(p.ID).(*PolicyCache), hash)
 	if !blocked {
 		return stackitem.NewBool(false)
 	}
@@ -359,7 +366,7 @@ func (p *Policy) unblockAccount(ic *interop.Context, args []stackitem.Item) stac
 // fee limit.
 func (p *Policy) CheckPolicy(d *dao.Simple, tx *transaction.Transaction) error {
 	for _, signer := range tx.Signers {
-		if _, isBlocked := p.isBlockedInternal(d, signer.Account); isBlocked {
+		if _, isBlocked := p.isBlockedInternal(d.GetROCache(p.ID).(*PolicyCache), signer.Account); isBlocked {
 			return fmt.Errorf("account %s is blocked", signer.Account.StringLE())
 		}
 	}

